include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h

```diff
@@ -0,0 +1,187 @@
+//===--- CanonicalOSSALifetime.h - Canonicalize OSSA lifetimes --*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// Canonicalize the copies and destroys of a single owned or guaranteed OSSA
+/// value.
+///
+/// This top-level API rewrites the extended lifetime of a SILValue:
+///
+///     void canonicalizeValueLifetime(SILValue def, CanonicalOSSALifetime&)
+///
+/// The extended lifetime transitively includes the uses of `def` itself along
+/// with the uses of any copies of `def`.
+///
+/// FIXME: Canonicalization currently bails out if any uses of the def has
+/// OperandOwnership::PointerEscape. Once project_box is protected by a borrow
+/// scope and mark_dependence is associated with an end_dependence,
+/// canonicalization will work everywhere as intended.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SILOPTIMIZER_UTILS_CANONICALOSSALIFETIME_H
+#define SWIFT_SILOPTIMIZER_UTILS_CANONICALOSSALIFETIME_H
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SetVector.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SILOptimizer/Utils/PrunedLiveness.h"
+
+namespace swift {
+
+struct CanonicalOSSALifetime;
+
+/// Top-Level API: rewrites copies and destroys within \p def's extended
+/// lifetime. \p lifetime caches transient analysis state across multiple calls
+/// and indicates whether any SILAnalyses must be invalidated.
+///
+/// Return false if the OSSA structure cannot be recognized (with a proper OSSA
+/// representation this will always return true).
+bool canonicalizeValueLifetime(SILValue def, CanonicalOSSALifetime &lifetime);
+
+/// Find the original definition of a potentially copied value.
+///
+/// This use-def walk must be consistent with the def-use walks performed within
+/// the CanonicalOSSALifetime implementation.
+inline SILValue getCanonicalCopiedDef(SILValue v) {
+  while (true) {
+    if (auto *copy = dyn_cast<CopyValueInst>(v)) {
```

Comment:
We already have a utility like this committed by Meghana. We shouldn't create these all over the place. Also I think you can do this cleaner with a while loop:

```
while (auto *copy = dyn_cast<CopyValueInst>(v))
  v = copy->getOperand();
return v;
```

Reply:
I won't use a separate utility here because, as the comment says, this implementation is coupled with the implementation inside `canonicalizeValueLifetime`. Other code that does not have such a dependence could use it though.

Reply:
Ok. I am fine with that.

Reply:
Can you hide it then and move it out of the global namespace?